### PR TITLE
Support custom renderer

### PR DESCRIPTION
**Pull request by paulfalgout**
Resolves https://github.com/marionettejs/backbone.marionette/issues/2474

Hopefully this allows for something like:
```js
import IDomRender from 'marionette-idom';
Marionette.View.setRenderer(IDomRender);
```

Todo:
- [x] Tests
- [x] Docs

### CODE REVIEW

**Comment by denar90** [comment]
I like it, so small change bring powerful API.

**Comment by scott-w** [comment]
Is there an advantage to doing it this way instead of allowing:

```javascript
import {Renderer} from 'backbone.marionette';
import IDomRenderer from 'marionette-idom';

Renderer.setRenderer(IDomRenderer);
```

**Comment by blikblum** [comment]
The `renderHtml` method is a nice addition allowing to more granular control 

> Is there an advantage to doing it this way instead of allowing:

Or as is today:

```
import {Renderer} from 'backbone.marionette';
import IDomRenderer from 'marionette-idom';

Renderer.render = IDomRenderer;
```

`Marionette.View.setRenderer(IDomRender);` does the same as using `Renderer.render`  but IMO, `View.setRenderer` is a better abstraction. We must consider deprecating Renderer.render, and remove Renderer altogether if this is applied

**Comment by paulfalgout** [comment]
Previous conversation https://github.com/marionettejs/backbone.marionette/pull/2911#issuecomment-197673522

Having the Render object is an extra function call, and I've hoping to potentially introduce other configurations via class functions:

`Marionette.View.setRenderer`
[`Marionette.View.setTemplateContext`]()
[`Marionette.View.setDomMixin`](https://github.com/marionettejs/backbone.marionette/issues/2164#issuecomment-197537666)(https://github.com/marionettejs/backbone.marionette/issues/3145#issuecomment-243033085)

Also looking back I realized this wasn't exactly correct.. I updated this so it is more granular on whichever view class you want to apply it to.

This makes me think that `renderHtml` should actually be `_renderHtml` and we would use the static functions for overriding it.

And that's mainly to keep people from having to really grasp prototypes.. setting `Marionette.View.prototype.renderHtml = function() { }` as on of the first things they might have to do when learning Marionette.

**Comment by paulfalgout** [comment]
I agree about deprecating the other renderer..  Can't figure out how to add the dep warning in the code though since it would have to know if you overrode it..  

Honestly I thought we might want to even extract the current renderer into a plugin if we could figure out optional dependencies.. it includes the templatecache.. which if you're overriding the renderer is a chunk of code that'll never be used.

**Comment by blikblum** [comment]
> This makes me think that renderHtml should actually be _renderHtml and we would use the static functions for overriding it.

IMO making `renderHtml` private is a step back.

Say i have a default renderer and want to customize a specific View.
With `renderHtml` public i would do:

```
MyView = Mn.View.extend({
  renderHtml: customRenderer
})
```

with `renderHtml` private:

```
MyView = Mn.View.extend({ })

MyView.setRenderer(customRenderer)
```
Both will have the same effect, but the former is consistent with current way of extending Mn classes and the later needs a second statement / call

Also, if the idea is to add also `Marionette.View.setTemplateContext`, to be consistent with the API setting a private method, `templateContext` would have to be made private also.

**Comment by paulfalgout** [comment]
I'm a little conflicted on this.  I made this a 2nd commit so it could be easily reverted, but I'm just not sure there's much use case for setting the renderer from view to view that often.  I'm personally still leaning towards having one way to do something with a smaller API.

I also don't think this is quite the same as how `setTemplateContext` will hopefully work.
I'm hoping `setTemplateContext` will allow for:

```js
Marionette.View.setTemplateContext({ BASE_URL: '/' });
//...
MyFormView.setTemplateContext({ current_user_name: currentUser.get('name') });
//...
var myBasicsForm = new MyFormView({
  template: _.template(`
       <form action="<%- BASE_URL %>basics">
        Hello <%- current_user_name %>,
        <input type="text" value="<%- phone_number_display %>">
      </form>
  `),
    templateContext: {
       phone_number_display() {
          return formatNumber(phone)
       }
    }
});
```

So part of the nicety of these setters is that you don't have to know what it's extending or how it works.  And you'll never have to ask which method to use.

Also going back to original example, it'd be nice if most of the renderer were like plugins.
```js
import IDomRender from 'marionette-idom';
Marionette.View.setRenderer(IDomRender);
```
Asking the user to do that requires 0 knowledge of the renderer, but asking the user to override the `renderHtml` function brings up one added layer of complexity.

I'm definitely not super sold on `_renderHtml`, but it'd be great to get a concensus @marionettejs/marionette-core

**Comment by scott-w** [comment]
I'd agree with your second commit and err on the side of making it private at first. We can always make it public in the next version if people want to use it.

**Comment by blikblum** [comment]
Thinking more, having only `setRenderer` as an API is a better option than exposing `renderHtml` method. So has my OK.

Just take a look at the comment about the implementation

**Comment by paulfalgout** [comment]
> comment about the implementation
@blikblum what comment specifically?

**Comment by blikblum** [comment]
> Why not this.protoype._renderHtml = renderer ?

File: src/view.js [hunk 1]
```diff
@@ -195,6 +200,11 @@ const View = Backbone.View.extend({
       .compact()
       .value();
   }
+}, {
+  // Sets the renderer for the Marionette.View class
+  setRenderer(renderer) {
+    _.extend(this.prototype, { _renderHtml: renderer });
```
Comment (paulfalgout) [inline src/view.js]: There it is :-)

**Comment by paulfalgout** [comment]
@marionettejs/marionette-core 👀

**Comment by paulfalgout** [comment]
@scott-w I'm nervous about modifying `Marionette.View.prototype` in the tests as there's only the one copy.
I'm not sure I 100% follow the list either.  And I wasn't sure about the standard renderer since that's currently tested in the renderer itself.. I suppose I can test that the standard rendered gets called if it isn't overridden...

**Comment by scott-w** [comment]
@paulfalgout Markdown didn't put my list together correctly. I meant to do the following:

```
I'm thinking we should get some tests in to cover a few potential edge-cases:

1. `View` has custom renderer, inherit from `View`:
  a. `View` uses custom renderer (covered)
  b. Inherited view uses custom renderer
2. Inherit from `View` without custom renderer:
  a. Inherited view uses custom renderer
  b. `View` uses standard renderer
```

I think you got the gist of it anyway - basically just making sure the right renderer gets used in the different cases.

**Comment by paulfalgout** [comment]
By 1b. do you mean the inherited view uses the same custom renderer as 1a?

**Comment by scott-w** [comment]
Yeah, I'm assuming that extending a view should share the same renderer?

**Comment by paulfalgout** [comment]
Ok so I have a `ViewClass` inherited from `Marionette.View` and a `SubViewClass` inherited from `ViewClass`.
I'm now checking to make sure:
`ViewClass` can have a custom renderer and if so that it doesn't affect `Marionette.View`, but it does affect the `SubViewClass`.
I'm also checking that the `SubViewClass` can have it's own custom renderer different than the class it inherited from.
Pretty sure that covers all of the scenarios.

**Comment by JSteunou** [comment]
:+1: for me too
